src/lib.rs

```diff
@@ -38,55 +31,245 @@ impl AndroidProperty {
+
+            Some(fn_ptr)
+        }
+
+        unsafe {
+            properties.read_callback_fn = load_fn(libc_so, "__system_property_read_callback")
+                .map(|raw| mem::transmute::<*const c_void, SystemPropertyReadCallbackFn>(raw));
+
+            properties.find_fn = load_fn(libc_so, "__system_property_find")
+                .map(|raw| mem::transmute::<*const c_void, SystemPropertyFindFn>(raw));
+
+            properties.set_fn = load_fn(libc_so, "__system_property_set")
+                .map(|raw| mem::transmute::<*const c_void, SystemPropertySetFn>(raw));
+
+            properties.for_each_fn = load_fn(libc_so, "__system_property_foreach")
+                .map(|raw| mem::transmute::<*const c_void, SystemPropertyForEachFn>(raw));
+
+            // Fallback for old versions of Android.
+            if properties.read_callback_fn.is_none() || properties.find_fn.is_none() {
+                properties.get_fn = load_fn(libc_so, "__system_property_get")
+                    .map(|raw| mem::transmute::<*const c_void, SystemPropertyGetFn>(raw));
+            }
+        }
+
+        properties
     }
-}
 
-impl fmt::Display for AndroidProperty {
-    // Output in format [<name>]: [<value>]
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        let mut property_info = self.property_info;
-        if property_info.is_null() {
-            property_info = plat_get_property_info(&self.name);
+    /// Retrieve a property with name `name`. Returns None if the operation fails.
+    pub fn get_property(&self, name: &str) -> Option<String> {
+        let cname = CString::new(name).unwrap();
+
+        self.get_property_impl(cname.as_ptr(), std::ptr::null())
+    }
+
+    fn get_property_impl(&self, cname: *const c_char, mut info: *const c_void) -> Option<String> {
+        // If available, use the recommended approach to accessing properties (Android L and onward).
+        if let (Some(find_fn), Some(read_callback_fn)) = (self.find_fn, self.read_callback_fn) {
+            if info.is_null() {
+                info = unsafe { (find_fn)(cname) };
+            }
+
+            if info.is_null() {
+                return None;
+            }
+
+            let mut result = Box::new(ValuePair {
+                name: String::new(),
+                value: String::new(),
+            });
+
+            unsafe {
+                (read_callback_fn)(info, property_callback, &mut *result);
+            }
+
+            return Some((result).value);
+        }
+
+        // Fall back to the older approach.
+        if let Some(get_fn) = self.get_fn {
+            const PROPERTY_VALUE_MAX: usize = 92;
+            let cvalue = CString::new(Vec::with_capacity(PROPERTY_VALUE_MAX)).unwrap();
+            let raw = cvalue.into_raw();
+            let ret = unsafe { (get_fn)(cname, raw) };
+            match ret {
+                len if len > 0 => unsafe { Some(String::from_raw_parts(raw as *mut u8, len as usize, PROPERTY_VALUE_MAX)) },
```

Comment:
Does Android guarantee that `get_fn` returns UTF-8? Constructing a `String` from bad UTF-8 is UB. I assume that's why `property_callback` calls `CStr`'s `to_str` method, which checks for valid UTF-8.